x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -313,9 +319,42 @@ private void markJobAsDeletingIfNotUsed(String jobId, ActionListener<Boolean> li
                                 + datafeedIds.iterator().next() + "] refers to it"));
                         return;
                     }
-                    jobConfigProvider.markJobAsDeleting(jobId, listener);
+                    cancelResetTaskIfExists(jobId, ActionListener.wrap(
+                        response -> jobConfigProvider.updateJobBlockReason(jobId, new Blocked(Blocked.Reason.DELETE, taskId), listener),
+                        listener::onFailure
+                    ));
                 },
                 listener::onFailure
         ));
     }
+
+    private void cancelResetTaskIfExists(String jobId, ActionListener<Boolean> listener) {
+        ActionListener<Job.Builder> jobListener = ActionListener.wrap(
+            jobBuilder -> {
+                Job job = jobBuilder.build();
+                if (job.getBlocked().getReason() == Blocked.Reason.RESET) {
+                    logger.info("[{}] Cancelling reset task [{}] because delete was requested", jobId, job.getBlocked().getTaskId());
+                    CancelTasksRequest cancelTasksRequest = new CancelTasksRequest();
+                    cancelTasksRequest.setReason("deleting job");
+                    cancelTasksRequest.setActions(ResetJobAction.NAME);
+                    cancelTasksRequest.setTaskId(job.getBlocked().getTaskId());
```

Comment:
I know that if `Blocked.Reason.RESET` then `job.getBlocked().getTaskId()` should never be null. But, if it is ever null, this would blow up. 

Same goes for any other tasks request. If we make a tasks request with a `null` taskId, this will go boom. 

I didn't see anything in this PR that struck me as the only time the taskID could be null is if the `Blocked.Reason.DELETE` and we don't ever check the task ID there. 

Just wanted to make sure we all knew this :)

Reply:
Yes, exactly.